core/messagetracer.go

```diff
@@ -0,0 +1,180 @@
+}
+
+// GetID returns the pluginID of the message source
+func (mts messageTracerSource) GetID() string {
+	return "core.MessageTracer"
+}
+
+// MessageTrace provide the MessageTrace() function. By default this function do nothing.
+var MessageTrace = func(msg *Message, pluginID string, comment string) {}
+
+// ActivateMessageTrace set a MessageTrace function to dump out the message trace
+func ActivateMessageTrace() {
+	MessageTrace = func(msg *Message, pluginID string, comment string) {
+		mt := messageTracer{
+			msg:      msg,
+			pluginID: pluginID,
+		}
+
+		mt.Dump(comment)
+	}
+}
+
+// DeactivateMessageTrace set a MessageTrace function to default
+// This method is necessary for unit testing
+func DeactivateMessageTrace() {
+	MessageTrace = func(msg *Message, pluginID string, comment string) {}
+}
+
+// Dump creates a messageDump struct for the message trace
+func (mt *messageTracer) Dump(comment string) {
+	if mt.msg.streamID == LogInternalStreamID || mt.msg.streamID == TraceInternalStreamID {
+		return
+	}
+
+	var err error
+	msgDump := mt.newMessageDump(mt.msg, comment)
+
+	if StreamRegistry.IsStreamRegistered(TraceInternalStreamID) {
+		err = mt.routeToTraceStream(msgDump)
+
+	} else {
+		err = mt.printMessageTrace(msgDump)
+	}
+
+	if err != nil {
+		logrus.Error(err)
+	}
+}
+
+func (mt *messageTracer) printMessageTrace(dump messageDump) error {
+	jsonData, err := json.MarshalIndent(dump, "", "\t")
+	if err == nil {
+		_, err = fmt.Println("\n", string(jsonData))
+	}
+
+	return err
+}
+
+func (mt *messageTracer) routeToTraceStream(dump messageDump) error {
+	jsonData, err := json.Marshal(dump)
+	if err != nil {
+		return err
+	}
+
+	traceMsg := NewMessage(messageTracerSource{}, jsonData, nil, TraceInternalStreamID)
+	traceRouter := StreamRegistry.GetRouterOrFallback(TraceInternalStreamID)
+
+	return traceRouter.Enqueue(traceMsg)
+}
+
+func (mt *messageTracer) newMessageDump(msg *Message, comment string) messageDump {
+	dump := messageDump{}
+	// set general trace info
+	dump.Step = comment
+	dump.PluginID = mt.pluginID
+
+	// prepare message payloads data
+	dump.Payload = msg.String()
+	if msg.orig != nil {
+		dump.OriginalPayload = string(msg.orig.payload)
+	}
+
+	// prepare message streams data
+	dump.Stream = mt.getStreamName(msg.streamID)
+	dump.PrevStream = mt.getStreamName(msg.prevStreamID)
+	dump.OrigStream = mt.getStreamName(msg.origStreamID)
+
+	// prepare meta data
+	if metadata := msg.TryGetMetadata(); metadata != nil {
+		dump.Metadata = map[string]string{}
+		for k, v := range metadata {
+			dump.Metadata[k] = string(v)
+		}
+	}
+
+	// set source data
+	if msg.source != nil {
+		dump.Source = msg.source.GetID()
+	}
+
+	//  set timestamp
+	dump.Timestamp = msg.timestamp
+
+	dump.FingerprintID = mt.createFingerPrintID(&dump)
+
+	return dump
+}
+
+func (mt *messageTracer) createFingerPrintID(dump *messageDump) uint32 {
+	hash := fnv.New32a()
+	hash.Write([]byte(fmt.Sprintf("%s/%s", dump.Source, dump.Timestamp)))
```

Comment:
you can append multiple values to a hash before sume.
Sprintf adds additional overhead which can be avoided.